rcl_yaml_param_parser/src/parser.c

```diff
@@ -1278,6 +1659,201 @@ static rcutils_ret_t parse_value(
+  yaml_scalar_style_t style = event.data.scalar.style;
+  const uint32_t line_num = ((uint32_t)(event.start_mark.line) + 1U);
+
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    value, "event argument has no value", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (style != YAML_SINGLE_QUOTED_SCALAR_STYLE &&
+    style != YAML_DOUBLE_QUOTED_SCALAR_STYLE &&
+    0U == val_size)
+  {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("No value at line %d", line_num);
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (NULL == params_st->descriptors[node_idx].parameter_descriptors) {
+    RCUTILS_SET_ERROR_MSG("Internal error: Invalid mem");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  rcl_param_descriptor_t * param_descriptor =
+    &(params_st->descriptors[node_idx].parameter_descriptors[parameter_idx]);
+
+  data_types_t val_type;
+  void * ret_val = get_value(value, style, &val_type, allocator);
+  if (NULL == ret_val) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error parsing value %s at line %d", value, line_num);
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (NULL == ns_tracker->parameter_ns) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Parameter assignment at line %d unallowed in parameter__descriptors", line_num);
+    return RCUTILS_RET_ERROR;
+  } else {
+    // If parsing a yaml value, then current parameter namespace must be parameter name
+    allocator.deallocate(params_st->descriptors[node_idx].parameter_names[parameter_idx],
+      allocator.state);
+    params_st->descriptors[node_idx].parameter_names[parameter_idx] =
+      rcutils_strdup(ns_tracker->parameter_ns, allocator);
+  }
+  if (NULL == param_descriptor->name) {
+    param_descriptor->name =
+      rcutils_strdup(params_st->descriptors[node_idx].parameter_names[parameter_idx], allocator);
+    rcl_node_params_descriptors_t * node_descriptors_st = &(params_st->descriptors[node_idx]);
+    node_descriptors_st->num_params++;
+  }
+  if (true == is_seq) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Sequences not supported for parameter descriptors at line %d", line_num);
+    return RCUTILS_RET_ERROR;
+  }
+
+  if (0 == strncmp("additional_constraints", ns_tracker->descriptor_key_ns,
+    strlen("additional_constraints")))
+  {
+    if (val_type == DATA_TYPE_STRING) {
+      param_descriptor->additional_constraints = (char *)ret_val;
+      if (NULL == param_descriptor->additional_constraints) {
+        RCUTILS_SAFE_FWRITE_TO_STDERR("Error allocating mem");
+        return RCUTILS_RET_BAD_ALLOC;
+      }
+    } else {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Value type 'string' expected at line %d for "
+        "parameter__descriptors key: additional_constraints",
+        line_num);
+      return RCUTILS_RET_ERROR;
+    }
+  } else if (0 == strncmp("type", ns_tracker->descriptor_key_ns, strlen("type"))) {
+    if (val_type == DATA_TYPE_INT64) {
+      param_descriptor->type = (uint8_t *)ret_val;
+      if (NULL == param_descriptor->type) {
```

Comment:
@bpwilcox here and everywhere else within this function, why do you check for `ret_val` nullity again? And why does the error message set reads as a memory allocation error?? Bad copy-paste?

Reply:
I'm not checking `ret_val` for nullity directly here, that is already done once when it is first assigned. Here I am following the same pattern as in the `parse_value`, where it checks the allocated descriptor struct field after being cast from `ret_val`. Am I misunderstanding the original code on that pattern?

Reply:
Hmm I'm not following you, I don't see such pattern. But even if there was, you're not allocating memory here when you cast (as the log message below suggests). It's a C-like cast, it won't result in NULL if it isn't NULL already. Or I completely missed the point of this code.

Reply:
I realized my mistake, you're correct. The check was required when allocating the arrays in `parse_value`, which I am not doing here.